### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by collinsugwu**
**what this pr does**
signup users
login users
create posts

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+class UsersController < ApplicationController
+  def new
+    @user = User.new
+  end
+
+  def create
```
Comment (Ispirett) [inline app/controllers/users_controller.rb]: This method is giving an error. You should follow rails  convention and create a user_params method
and then pass this method to `User.new `method

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    @user = User.new(params[:user])
+    if @user.save
+      session[:user_id] = @user.id
+      redirect_to root_url, notice: 'Thank you for signing up!'
```
Comment (Ispirett) [inline app/controllers/users_controller.rb]: root url_l is giving and error you can just user `'/'` instead

File: config/routes.rb [hunk 3]
```diff
@@ -1,3 +1,8 @@
 Rails.application.routes.draw do
+  get 'sign-up', to: 'users#new', as: 'sign_up'
+  get 'sessions/new'
+  post '/sign-in', to: 'session#create', as: 'sign_in'
```
Comment (Ispirett) [inline config/routes.rb]: you need a get method for user sign in. I can't sign-in

File: app/views/sessions/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+<h1>Sessions#new</h1>
```
Comment (Ispirett) [inline app/views/sessions/new.html.erb]: You need a form here to sign in

File: config/routes.rb [hunk 5]
```diff
@@ -1,3 +1,8 @@
 Rails.application.routes.draw do
+  get 'sign-up', to: 'users#new', as: 'sign_up'
+  get 'sessions/new'
```
Comment (Ispirett) [inline config/routes.rb]: This is no url_paths for posts . Post can't be accessed

File: app/views/posts/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<div class="container mt-5">
+  <div class="card">
+  <div class="card-body">
+    <h5 class="card-title"><%= @post.title %></h5>
+    <p class="card-text"><%= @post.body %></p>
+  </div>
+</div>
+</div>
```
[the remark this answers is not in the document]
Comment (maelfosso) [reply app/views/posts/show.html.erb]: Same problem !!

File: app/controllers/posts_controller.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+class PostsController < ApplicationController
+  before_action :current_user, only: %i[new create]
```
[the remark this answers is not in the document]
Comment (maelfosso) [reply app/controllers/posts_controller.rb]: Same problem !!! I can directly access to `posts/new` when the user it's not logged in and I have an error

Create an another method which will use `current_user` and **redirect** to login if currrent it's nil

File: app/helpers/sessions_helper.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+module SessionsHelper
+  def sign_in(user)
+    token = SecureRandom.urlsafe_base64
+    user.remember_token = Digest::SHA1.hexdigest(token.to_s)
+    user.save
```
[the remark this answers is not in the document]
Comment (maelfosso) [reply app/helpers/sessions_helper.rb]: You can write that in one line

File: app/helpers/sessions_helper.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    cookies.permanent[:token] = {
+      value: token,
+      domain: '127.0.0.1:3000',
+      expires: 1.day.from_now.utc
+    }
+    assign_current_user(user)
+  end
+
+  def current_user
+    @user ||= User.find(session[:user_id]) if session[:user_id]
```
[the remark this answers is not in the document]
Comment (maelfosso) [reply app/helpers/sessions_helper.rb]: No need to use `session` here.
Use the remember token saved in cookies.permanent

File: app/helpers/sessions_helper.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  end
+
+  def assign_current_user(user)
+    session[:user_id] = user.id
```
Comment (maelfosso) [inline app/helpers/sessions_helper.rb]: No need to use `session`  !!!

**Pull request merged**